hook!-full-game/addons/level_design_dock/plugin.gd

```diff
@@ -0,0 +1,111 @@
+tool
+extends EditorPlugin
+
+const INTERFACE_SCENE: = preload("res://addons/level_design_dock/LevelDesignDock.tscn")
+const LevelDesignDock: = preload("res://addons/level_design_dock/LevelDesignDock.gd")
+const PackedSceneButton: = preload("res://addons/level_design_dock/PackedSceneButton.gd")
+
+var _interface: LevelDesignDock
+var _current_instance: Node2D setget _set_current_instance
+var _current_parent: Node setget _set_current_parent
+var _current_scene: PackedScene setget _set_current_scene
+
+func _enter_tree() -> void:
+	_interface = INTERFACE_SCENE.instance()
+	add_control_to_dock(EditorPlugin.DOCK_SLOT_RIGHT_UR, _interface)
+	_interface.connect("packed_scene_button_created", self, "_on_Interface_PackedScene_Button_Created")
+
+
+func _exit_tree() -> void:
+	_interface.queue_free()
+
+
+func handles(object: Object) -> bool:
+	var can_handle = object is Node
+	if can_handle:
+		_set_current_parent(object)
+	return can_handle
+
+
+func forward_canvas_gui_input(event: InputEvent) -> bool:
+	var forward: = false
+	if event is InputEventMouseMotion:
+		_move_instance()
+		forward = true
+	elif event is InputEventMouseButton:
+		if event.button_index == BUTTON_LEFT:
+			forward = true
+			if event.is_pressed():
+				_create_instance()
+			else:
+				_clear_current_instance()
+	if event is InputEventKey:
+		if event.scancode == KEY_ESCAPE:
+			_clear_current_instance()
+			_clear_current_parent()
+			_clear_current_scene()
+			forward = true
+	return forward
+
+
+func _move_instance() -> void:
+	var can_drag = Input.is_mouse_button_pressed(BUTTON_LEFT) and _current_instance
+	if can_drag:
+		_current_instance.global_position = _current_instance.get_global_mouse_position()
+
+
+func _on_Interface_PackedScene_Button_Created(button: PackedSceneButton) -> void:
+	button.connect("toggled", self, "_on_PackedSceneButton_button_toggled", [button])
+
+
+func _on_PackedSceneButton_button_toggled(button_pressed: bool, button: PackedSceneButton):
+	if button_pressed:
+		_set_current_scene(button.packed_scene)
+	else:
+		_clear_current_scene()
+
+
+func _can_instance() -> bool:
+	var can_instance = false
+	if _current_parent and _current_scene:
+		can_instance = true
+	return can_instance
```

Comment:
You can write `return _current_parent and _current_scene` directly

Reply:
Isn't it better to have the condition/value explicitly communicated through code?

`return _can_current_parent and _current_scene` gives a lot of space for misinterpretations and hide information from the reader, it's a very implicit way to do approach it, not to mention that further, we may have new conditions that change the value returned by this function, meaning something like `var can_instance: bool` would have to be added anyway.

Reply:
It doesn't leave room for misinterpretations to me. The function already is `_can_instance`, it returns a boolean value. `_current_parent and _current_scene` reads as `is there a current parent and a scene?`.
If the function has to become complex enough you need a bool variable someday, you can take 10 seconds to add it. You could also have:

```gdscript
var can_instance := _current_parent and _current_scene
return can_instance
```

I don't mind, even though to me this is clear:
```gdscript
func _can_instance() -> bool:
    return _current_parent and _current_scene
```